File: src/service/github.service.js
import axios from "axios";

export default {
  montarUrl(busca) {
    if (!busca) {
      throw new Error("Campo de pesquisa não pode estar nulo");
    }

    const URL = "https://api.github.com/search/repositories";
    const SORT = "sort=stars";
    const PAGE = "page=0";
    const SIZE = "per_page=5";
    const Q = `q=language:${busca}`

    return `${URL}?${SORT}&${PAGE}&${SIZE}&${Q}`;
  },

  buscarRepositorios(url) {
    console.log(url)
    return axios.get(url);
  }
}

File: lib/highlight_repo/services/github_service.ex
defmodule HighlightRepo.Services.GithubService do
  @moduledoc """
  GithubService
  """
  @github_repos_api "https://api.github.com/search/repositories"

  defp call_github_api(language) do
    url = @github_repos_api <> "?q=language:#{language}&sort=stars&order=desc&page=1&per_page=10"

    {:ok, response} = HTTPoison.get(url)

    case response.status_code == 200 do
      true -> Jason.decode(response.body)
      false -> {:error, "Github api not available at the moment, please try again later"}
    end
  end

  @spec get_repos_by_language(any) :: any
  def get_repos_by_language(language) do
    case call_github_api(language) do
      {:ok, repos} ->
        {:ok, repos["items"]}

      {:error, message} ->
        {:error, message}
    end
  end
end
